ouroboros-network/src/Ouroboros/Network/Mux.hs

```diff
@@ -139,3 +159,85 @@ muxChannel pmss mid md w cnt =
            then pure Nothing
            else return $ Just blob
 
+muxBearerNew :: (MonadSTM m, Ord ptcl, Enum ptcl, Bounded ptcl)
+             => (MuxSDU ptcl -> m (Time m)) -- Write action
+             -> m (MuxSDU ptcl, Time m)     -- Read action
+             -> m Word16                    -- SDU size action
+             -> m ()                        -- Close action
+             -> m (MuxBearer ptcl m)
+muxBearerNew wfn rwn sfn cfn = do
+    st <- atomically $ newTVar Larval
+    return $ MuxBearer wfn rwn sfn cfn st
+
+muxBearerSetState :: (MonadSTM m, Ord ptcl, Enum ptcl, Bounded ptcl)
+                  => MuxBearer ptcl m
+                  -> MuxBearerState
+                  -> m ()
+muxBearerSetState bearer newState = atomically $ writeTVar (state bearer) newState
+
+
+muxClient :: (MonadAsync m, MonadFork m, MonadSay m, MonadSTM m, MonadThrow m,
+            Ord ptcl, Enum ptcl, Bounded ptcl, HasCallStack)
+        => [(Version, MiniProtocolDescriptions ptcl m)]
+        -> MuxBearer ptcl m
+        -> m (Version, MiniProtocolDescriptions ptcl m)
+muxClient verMpds bearer = do
+    let versions = map fst verMpds
+        msg = MsgInitReq versions
+        blob = toLazyByteString $ encodeCtrlMsg msg
+        sdu = MuxSDU (RemoteClockModel 0) Muxcontrol ModeInitiator (fromIntegral $ BL.length blob) blob
+
+    void $ write bearer sdu
+    (rsp, _) <- Ouroboros.Network.Mux.Types.read bearer
+    if msId rsp /= Muxcontrol || msMode rsp /= ModeResponder
```

Comment:
This `if` is always `True` :), maybe you intended to have `&&` instead of `||`.

Reply:
I think that the ìf` statement is correct, the client expects a `MuxControl` reply with the `ModeResponder` set, anything else is an error.

Reply:
You're right, I read it wrong as if it would be `rsp /= Muxcontrol || rsp /= ModeResponder`.